File: pr_comments/fetcher.py
"""Fetcher for GitHub PR comments using gh CLI."""

import json
import subprocess
from typing import Any


class GitHubAPIError(Exception):
    """Raised when GitHub API call fails."""

    pass


def fetch_pr_comments(owner: str, repo: str, pr_number: int) -> list[dict[str, Any]]:
    """Fetch PR comments using the gh CLI.

    Args:
        owner: Repository owner (e.g., 'ROKT').
        repo: Repository name (e.g., 'canal').
        pr_number: Pull request number.

    Returns:
        List of comment dicts from the GitHub API.

    Raises:
        GitHubAPIError: If the API call fails.
    """
    endpoint = f"repos/{owner}/{repo}/pulls/{pr_number}/comments"

    try:
        result = subprocess.run(
            ["gh", "api", endpoint],
            capture_output=True,
            text=True,
            check=True,
        )
        return json.loads(result.stdout)
    except subprocess.CalledProcessError as e:
        raise GitHubAPIError(f"Failed to fetch PR comments: {e.stderr}") from e
    except json.JSONDecodeError as e:
        raise GitHubAPIError(f"Failed to parse API response: {e}") from e


def fetch_pr_review_comments(owner: str, repo: str, pr_number: int) -> list[dict[str, Any]]:
    """Fetch PR review comments (issue comments) using the gh CLI.

    These are general comments on the PR, not tied to specific code lines.

    Args:
        owner: Repository owner.
        repo: Repository name.
        pr_number: Pull request number.

    Returns:
        List of comment dicts from the GitHub API.

    Raises:
        GitHubAPIError: If the API call fails.
    """
    endpoint = f"repos/{owner}/{repo}/issues/{pr_number}/comments"

    try:
        result = subprocess.run(
            ["gh", "api", endpoint],
            capture_output=True,
            text=True,
            check=True,
        )
        return json.loads(result.stdout)
    except subprocess.CalledProcessError as e:
        raise GitHubAPIError(f"Failed to fetch PR review comments: {e.stderr}") from e
    except json.JSONDecodeError as e:
        raise GitHubAPIError(f"Failed to parse API response: {e}") from e


def fetch_pr_info(owner: str, repo: str, pr_number: int) -> dict[str, Any]:
    """Fetch basic PR information.

    Args:
        owner: Repository owner.
        repo: Repository name.
        pr_number: Pull request number.

    Returns:
        PR info dict from the GitHub API.

    Raises:
        GitHubAPIError: If the API call fails.
    """
    endpoint = f"repos/{owner}/{repo}/pulls/{pr_number}"

    try:
        result = subprocess.run(
            ["gh", "api", endpoint],
            capture_output=True,
            text=True,
            check=True,
        )
        return json.loads(result.stdout)
    except subprocess.CalledProcessError as e:
        raise GitHubAPIError(f"Failed to fetch PR info: {e.stderr}") from e
    except json.JSONDecodeError as e:
        raise GitHubAPIError(f"Failed to parse API response: {e}") from e

File: go-pr-comments/fetcher.go
package main

import (
	"encoding/json"
	"fmt"
	"os/exec"
)

// GitHubAPIError represents an error from the GitHub API.
type GitHubAPIError struct {
	Message string
}

func (e *GitHubAPIError) Error() string {
	return e.Message
}

// RawComment represents the raw JSON structure from GitHub API.
type RawComment struct {
	ID                int    `json:"id"`
	Path              string `json:"path"`
	Line              *int   `json:"line"`
	OriginalLine      *int   `json:"original_line"`
	StartLine         *int   `json:"start_line"`
	OriginalStartLine *int   `json:"original_start_line"`
	Body              string `json:"body"`
	CreatedAt         string `json:"created_at"`
	UpdatedAt         string `json:"updated_at"`
	DiffHunk          string `json:"diff_hunk"`
	HTMLURL           string `json:"html_url"`
	User              struct {
		Login string `json:"login"`
	} `json:"user"`
}

// PRInfo represents basic PR information from GitHub API.
type PRInfo struct {
	HTMLURL string `json:"html_url"`
	Title   string `json:"title"`
}

// FetchPRComments fetches PR comments using the gh CLI.
func FetchPRComments(owner, repo string, prNumber int) ([]RawComment, error) {
	endpoint := fmt.Sprintf("repos/%s/%s/pulls/%d/comments", owner, repo, prNumber)

	cmd := exec.Command("gh", "api", endpoint)
	output, err := cmd.Output()
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			return nil, &GitHubAPIError{
				Message: fmt.Sprintf("Failed to fetch PR comments: %s", string(exitErr.Stderr)),
			}
		}
		return nil, &GitHubAPIError{
			Message: fmt.Sprintf("Failed to fetch PR comments: %v", err),
		}
	}

	var comments []RawComment
	if err := json.Unmarshal(output, &comments); err != nil {
		return nil, &GitHubAPIError{
			Message: fmt.Sprintf("Failed to parse API response: %v", err),
		}
	}

	return comments, nil
}

// FetchPRInfo fetches basic PR information.
func FetchPRInfo(owner, repo string, prNumber int) (*PRInfo, error) {
	endpoint := fmt.Sprintf("repos/%s/%s/pulls/%d", owner, repo, prNumber)

	cmd := exec.Command("gh", "api", endpoint)
	output, err := cmd.Output()
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			return nil, &GitHubAPIError{
				Message: fmt.Sprintf("Failed to fetch PR info: %s", string(exitErr.Stderr)),
			}
		}
		return nil, &GitHubAPIError{
			Message: fmt.Sprintf("Failed to fetch PR info: %v", err),
		}
	}

	var info PRInfo
	if err := json.Unmarshal(output, &info); err != nil {
		return nil, &GitHubAPIError{
			Message: fmt.Sprintf("Failed to parse API response: %v", err),
		}
	}

	return &info, nil
}
